### PR TITLE
Merge Progress Feature and Data Feature Backend

### PR DESCRIPTION
**Pull request by ike-elue**
Add:
- JsonGoalStep; object that can be stored as a json in the Data feature and has all the necessary information to make a normal Goal Step (for the MVP, this specifically targets marathons).

Important Updates:
- ProgressModel; moved around a lot of code for readability, implemented builder pattern, and established connections with the Data feature (see toJson() and getLastSession() in file).
- Exercise; changed "better than" and "equal to" logic (see design doc for reasoning)
- GoalStep; has conversion between itself and JsonGoalStep.
- Session; has conversion between itself and MarathonSession.
- CreateProfileServlet; added goal steps json to user entity.
- DataHandler; added Goal Steps property and getter/setter for goal steps (used Text instead of String because datastore caps out at 70 characters for String). 

Removed
- Data mock; I now have the Data feature for those functions

**TL;DR**: Essentially, this pr does the same logic as before, but with the Data feature in mind.

### CODE REVIEW

File: capstone/fitness/src/main/java/com/google/sps/progress/JsonGoalStep.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+package com.google.sps.progress;
+
+import com.google.sps.fit.*;
+import com.google.sps.fit.Exercise.SetType;
+import com.google.sps.util.*;
+import java.util.*;
+
+/* Json version of progress marker. */
+public class JsonGoalStep {
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/progress/JsonGoalStep.java:23]: Can we add methods in GoalStep for conversion between json and pojo instead of have a new class JsonGoalStep or even better use a lib like Jackson API?

File: capstone/fitness/src/main/java/com/google/sps/servlets/CreateProfileServlet.java [hunk 2]
```diff
@@ -19,10 +21,15 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import com.google.appengine.api.datastore.FetchOptions;
+import com.google.gson.Gson;
+import java.util.*;
 
 @WebServlet("/create-profile")
 public class CreateProfileServlet extends HttpServlet {
   
+  private final String exerciseName = "Running";
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/servlets/CreateProfileServlet.java]: make the constants static

File: capstone/fitness/src/main/java/com/google/sps/util/DataHandler.java [hunk 3]
```diff
@@ -123,6 +116,23 @@ public static String getData(String property, Entity user) {
     return data; 
   }
 
+  public static String getGoalSteps() {
```
Comment (lupengscu) [inline capstone/fitness/src/main/java/com/google/sps/util/DataHandler.java:119]: would it be better to return a list of GoalStep obj? But not sure how this method is going to be used.

**Review by lupengscu — APPROVED** [review]

File: capstone/fitness/src/main/java/com/google/sps/util/DataHandler.java (hunk 3 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/util/DataHandler.java:119]: I was planning on using this returned json in the setJsonGoalSteps of ProgressModel Builder. I didn't want to take away from the normal return types of the DataHandler (everything returned from the file is a json or datastore related entity) so I thought returning the json made sense. However, I could just move the logic from the Builder method to this function.

File: capstone/fitness/src/main/java/com/google/sps/servlets/CreateProfileServlet.java (hunk 2 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/servlets/CreateProfileServlet.java]: done

File: capstone/fitness/src/main/java/com/google/sps/progress/JsonGoalStep.java (hunk 1 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/progress/JsonGoalStep.java:23]: I looked into the Jackson API and it seems like it also uses a pojo to act as an intermediate with reading and writing. Also it seems like it has to write to a local file/output stream to produce json objects.

I mostly used the pojo because Gson made conversion really simple to and from the GoalStep (see toJson and setJsonGoalSteps in ProgressModel). 

Also, what kind of methods did you have in mind for the conversion between json and pojo?

**Review by GabrielGonzalezDiaz — APPROVED** [review]

File: capstone/fitness/src/main/java/com/google/sps/fit/Exercise.java [hunk 4]
```diff
@@ -113,82 +117,85 @@ public Exercise build() {
   private Exercise(){}
 
   /**
-   * Returns true if average of Exercise values are better than the given Exercise's average 
-   * values in each type.
-   * Note: Uses average to ignore set count.
+   * Returns true if Exercise's values are better than or overall equal to the given Exercise's 
+   * values in each type. Uses average if values should decrement, uses summation if not. The equal
```
Comment (sliang3) [inline capstone/fitness/src/main/java/com/google/sps/fit/Exercise.java:121]: Not blocking for this PR, but it doesn't seem immediately obvious why decreasing values should be average but the opposite should be summed. Maybe this can be something put into SetType for the future.

File: capstone/fitness/src/main/java/com/google/sps/fit/Exercise.java [hunk 5]
```diff
@@ -113,82 +117,85 @@ public Exercise build() {
...
+    boolean betterThanAtLeastOnce = false;
     for(SetType type : setValues.keySet()) {
-      float srcAvg = avg(getSetValues(type));
-      float compAvg = avg(exercise.getSetValues(type));
-      if(!type.betterThan(srcAvg, compAvg)) {
-        return false;
-      }
+      boolean betterThan = betterThan(exercise, type).orElse(false);
+      betterThanAtLeastOnce |= betterThan;
+      betterThanOverall &= betterThan || equalTo(exercise, type).orElse(false);
```
Comment (sliang3) [inline capstone/fitness/src/main/java/com/google/sps/fit/Exercise.java]: It seems like you can just early return false here if this expression becomes false?

**Review by sliang3 — APPROVED** [review]

File: capstone/fitness/src/main/java/com/google/sps/fit/Exercise.java (hunk 4 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/fit/Exercise.java:121]: Yeah my bad, the math just kinda seemed to work out for the example edge cases I gave myself. For example, if I was comparing a 600 sec duration to two 300 sec durations, I would say the two 300s are better in the decrement case because they are smaller numbers. However, if I use summation then the algorithm would claim they are equal even though doing one set of 600 sec for something would be significantly easier than two sets of 300 secs (e.g. Running). I'll probably move this logic to SetType though because it does seem to clutter the functions a lot.

**TL;DR** - The math just kinda works (see design doc for examples), but I'll eventually move the logic to SetType someday.

File: capstone/fitness/src/main/java/com/google/sps/fit/Exercise.java (hunk 5 above)
Comment (ike-elue) [reply capstone/fitness/src/main/java/com/google/sps/fit/Exercise.java]: done
